File: autonav_ws/src/autonav_shared/src/node.cpp
#include "autonav_shared/node.hpp"
#include <unistd.h>

namespace AutoNav
{
    Node::Node(const std::string & node_name, const bool has_ownership) : rclcpp::Node(node_name)
    {
        // Set ownership
        this->has_ownership = has_ownership;

        // Setup our device state
        system_state = AutoNav::SystemState::DISABLED;
        device_states.insert_or_assign(node_name, AutoNav::DeviceState::OFF);

        // TODO: Setup all relevant publishers, subscribers, services, clients, etc
        system_state_sub = this->create_subscription<autonav_msgs::msg::SystemState>("/autonav/shared/system", 1, std::bind(&Node::system_state_callback, this, std::placeholders::_1));
        device_state_sub = this->create_subscription<autonav_msgs::msg::DeviceState>("/autonav/shared/device", 1, std::bind(&Node::device_state_callback, this, std::placeholders::_1));
        configuration_broadcast_sub = this->create_subscription<autonav_msgs::msg::ConfigurationBroadcast>("/autonav/shared/config/requests", 1, std::bind(&Node::config_broadcast_callback, this, std::placeholders::_1));
        configuration_update_sub = this->create_subscription<autonav_msgs::msg::ConfigurationUpdate>("/autonav/shared/config/updates", 1, std::bind(&Node::config_updated_callback, this, std::placeholders::_1));

        performance_pub = this->create_publisher<autonav_msgs::msg::Performance>("/autonav/shared/performance", 10);
        log_pub = this->create_publisher<autonav_msgs::msg::Log>("/autonav/shared/log", 10);
        device_state_pub = this->create_publisher<autonav_msgs::msg::DeviceState>("/autonav/shared/device", 10);
        system_state_pub = this->create_publisher<autonav_msgs::msg::SystemState>("/autonav/shared/system", 10);
        configuration_broadcast_pub = this->create_publisher<autonav_msgs::msg::ConfigurationBroadcast>("/autonav/shared/config/requests", 10);
        configuration_update_pub = this->create_publisher<autonav_msgs::msg::ConfigurationUpdate>("/autonav/shared/config/updates", 10);
    }

    Node::~Node()
    {
        // TODO: Cleanup stuff as required
    }

    void Node::config_updated_callback(const autonav_msgs::msg::ConfigurationUpdate::SharedPtr msg)
    {
        if (msg->device == get_name())
        {
            // log("Received update on our own configuration", Logging::LogLevel::DEBUG);
            json old_config = this->_config;
            // this->_config = json::parse(msg->json);
            apply_config(json::parse(msg->json));
            this->on_config_updated(old_config, this->_config);
        } else {
            // log("Received update on " + msg->device + "'s configuration", Logging::LogLevel::DEBUG);
        }

        other_cfgs.insert_or_assign(msg->device, json::parse(msg->json));
    }

    void Node::config_broadcast_callback(const autonav_msgs::msg::ConfigurationBroadcast::SharedPtr msg)
    {
        if (msg->target_nodes.size() == 0 || std::find(msg->target_nodes.begin(), msg->target_nodes.end(), get_name()) != msg->target_nodes.end())
        {
            log("Received request for our own configuration to be broadcasted", Logging::LogLevel::DEBUG);
            broadcast_config();
        };
    }

    void Node::broadcast_config()
    {
        autonav_msgs::msg::ConfigurationUpdate msg;
        msg.device = get_name();
        msg.json = _config.dump();
        configuration_update_pub->publish(msg);
    }

    void Node::request_all_configs()
    {
        autonav_msgs::msg::ConfigurationBroadcast msg;
        configuration_broadcast_pub->publish(msg);
    }

    void Node::request_config(const std::string & device)
    {
        autonav_msgs::msg::ConfigurationBroadcast msg;
        msg.target_nodes.push_back(device);
        configuration_broadcast_pub->publish(msg);
    }

    void Node::write_config(const json & config)
    {
        this->_config = config;
        broadcast_config();
    }

    void Node::perf_start(const std::string & name)
    {
        start_times.insert_or_assign(name, std::chrono::high_resolution_clock::now());
    }

    void Node::perf_stop(const std::string & name, const bool print_to_console)
    {
        if (start_times.find(name) == start_times.end())
        {
            log("Performance timer " + name + " not found", Logging::LogLevel::ERROR);
            return;
        }

        // Calculate the duration
        auto end_time = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_times.at(name));

        // Publish the performance data
        auto performance_msg = autonav_msgs::msg::Performance();
        performance_msg.name = name;
        performance_msg.elapsed = duration.count();
        performance_pub->publish(performance_msg);

        // Log the performance data
        if (print_to_console)
        {
            log("Performance timer " + name + " took " + std::to_string(duration.count()) + "ms", Logging::LogLevel::DEBUG);
        }

        // Remove the timer
        start_times.erase(name);
    }

    void Node::set_device_state(const std::string & device, AutoNav::DeviceState state)
    {
        // Publish the update
        device_states.insert_or_assign(device, state);
        autonav_msgs::msg::DeviceState msg;
        msg.device = device;
        msg.state = static_cast<uint8_t>(state);
        device_state_pub->publish(msg);
    }

    void Node::set_system_state(AutoNav::SystemState state, bool has_mobility)
    {
        // Publish the update
        autonav_msgs::msg::SystemState msg;
        msg.state = static_cast<uint8_t>(state);
        msg.mobility = has_mobility;
        system_state_pub->publish(msg);
        system_state = state;
        this->has_mobility = has_mobility;
    }

    void Node::system_state_callback(const autonav_msgs::msg::SystemState::SharedPtr msg)
    {
        AutoNav::SystemState old_state = system_state;
        bool old_mobility = has_mobility;

        system_state = static_cast<AutoNav::SystemState>(msg->state);
        has_mobility = msg->mobility;

        if (system_state != old_state)
        {
            on_system_state_updated(old_state, system_state);
        }

        if (has_mobility != old_mobility)
        {
            on_mobility_updated(old_mobility, has_mobility);
        }

        if (system_state == SystemState::SHUTDOWN)
        {
            kill(getpid(), SIGKILL);
        }
    }

    void Node::device_state_callback(const autonav_msgs::msg::DeviceState::SharedPtr msg)
    {
        if (msg->device != get_name())
        {
            device_states.insert_or_assign(msg->device, static_cast<AutoNav::DeviceState>(msg->state));
            return;
        }
        
        bool is_warming_up = device_states.at(msg->device) == AutoNav::DeviceState::OFF && static_cast<AutoNav::DeviceState>(msg->state) == AutoNav::DeviceState::WARMING;
        if (is_warming_up)
        {
            log("Received warming up state from " + msg->device, Logging::LogLevel::DEBUG);
            init();
        }
        device_states.insert_or_assign(msg->device, static_cast<AutoNav::DeviceState>(msg->state));
    }

    // TODO: Log to file
    void Node::log(const std::string & message, Logging::LogLevel level, const std::string & function_caller, int line)
    {        
        // Current time/date
        auto now = std::chrono::system_clock::now();
        std::time_t now_time = std::chrono::system_clock::to_time_t(now);
        std::tm now_tm = *std::localtime(&now_time);
        std::chrono::milliseconds ms = std::chrono::duration_cast<std::chrono::milliseconds>(now.time_since_epoch()) % 1000;
        
        // Log level
        std::string log_level = AutoNav::Logging::LOG_LEVEL_NAMES.at(level);
        if (log_level.length() < 5)
        {
            log_level += std::string(5 - log_level.length(), ' ');
        }

        // Log to topic
        autonav_msgs::msg::Log log_msg;
        log_msg.timestamp = now_time;
        log_msg.line_number = line;
        log_msg.function_caller = function_caller;
        log_msg.message = message;
        log_msg.node = get_name();
        log_msg.level = static_cast<uint8_t>(level);
        log_pub->publish(log_msg);

        switch (level)
        {
            case Logging::LogLevel::DEBUG:
                std::cout << termcolor::color<99, 150, 79> << std::put_time(&now_tm, "%Y-%m-%d %H:%M:%S") << ":" << std::setfill('0') << std::setw(3) << ms.count() 
                    << termcolor::white << " | " 
                    << termcolor::color<61, 117, 157> << log_level 
                    << termcolor::white << " | " 
                    << termcolor::color<90, 60, 146> << get_name() 
                    << termcolor::white << ":" 
                    << termcolor::color<90, 60, 146> << function_caller 
                    << termcolor::white << ":" 
                    << termcolor::color<90, 60, 146> << line 
                    << termcolor::white << " - " 
                    << termcolor::color<61, 117, 157> << message << termcolor::reset << std::endl;
                break;
            case Logging::LogLevel::INFO:
                std::cout << termcolor::color<99, 150, 79> << std::put_time(&now_tm, "%Y-%m-%d %H:%M:%S") << ":" << std::setfill('0') << std::setw(3) << ms.count() 
                    << termcolor::white << " | " 
                    << termcolor::color<255, 255, 255> << log_level 
                    << termcolor::white << " | " 
                    << termcolor::color<90, 60, 146> << get_name() 
                    << termcolor::white << ":" 
                    << termcolor::color<90, 60, 146> << function_caller 
                    << termcolor::white << ":" 
                    << termcolor::color<90, 60, 146> << line 
                    << termcolor::white << " - " 
                    << termcolor::color<255, 255, 255> << message << termcolor::reset << std::endl;
                break;
            case Logging::LogLevel::WARN:
                std::cout << termcolor::color<99, 150, 79> << std::put_time(&now_tm, "%Y-%m-%d %H:%M:%S") << ":" << std::setfill('0') << std::setw(3) << ms.count() 
                    << termcolor::white << " | " 
                    << termcolor::color<226, 174, 47> << log_level 
                    << termcolor::white << " | " 
                    << termcolor::color<90, 60, 146> << get_name() 
                    << termcolor::white << ":" 
                    << termcolor::color<90, 60, 146> << function_caller 
                    << termcolor::white << ":" 
                    << termcolor::color<90, 60, 146> << line 
                    << termcolor::white << " - " 
                    << termcolor::color<226, 174, 47> << message << termcolor::reset << std::endl;
                break;
            case Logging::LogLevel::ERROR:
                std::cout << termcolor::color<99, 150, 79> << std::put_time(&now_tm, "%Y-%m-%d %H:%M:%S") << ":" << std::setfill('0') << std::setw(3) << ms.count() 
                    << termcolor::white << " | " 
                    << termcolor::color<195, 59, 91> << log_level 
                    << termcolor::white << " | " 
                    << termcolor::color<90, 60, 146> << get_name() 
                    << termcolor::white << ":" 
                    << termcolor::color<90, 60, 146> << function_caller 
                    << termcolor::white << ":" 
                    << termcolor::color<90, 60, 146> << line 
                    << termcolor::white << " - " 
                    << termcolor::color<195, 59, 91> << message << termcolor::reset << std::endl;
                break;
            case Logging::LogLevel::FATAL: // rgb(207, 62, 99)
                std::cout << termcolor::color<99, 150, 79> << std::put_time(&now_tm, "%Y-%m-%d %H:%M:%S") << ":" << std::setfill('0') << std::setw(3) << ms.count() 
                    << termcolor::white << " | " 
                    << termcolor::white << termcolor::on_color<207, 62, 99> << log_level 
                    << termcolor::reset << termcolor::white << " | " 
                    << termcolor::color<90, 60, 146> << get_name() 
                    << termcolor::white << ":" 
                    << termcolor::color<90, 60, 146> << function_caller 
                    << termcolor::white << ":" 
                    << termcolor::color<90, 60, 146> << line 
                    << termcolor::white << " - " 
                    << termcolor::white << termcolor::on_color<207, 62, 99> << message << termcolor::reset << std::endl;
                break;
        }
    }
}

File: autonav_ws/src/autonav_shared/autonav_shared/node.py
import rclpy
from rclpy.node import Node as RclpyNode
from autonav_shared.types import DeviceState, LogLevel, SystemState
from autonav_msgs.msg import SystemState as SystemStateMsg, DeviceState as DeviceStateMsg, Performance, Log, ConfigurationBroadcast, ConfigurationUpdate
from rclpy.executors import MultiThreadedExecutor
import sty
import time
import inspect
import json


class Node(RclpyNode):
    def __init__(self, name: str) -> None:
        super().__init__(name)
        
        # Setup our device state
        self.system_state = SystemState.DISABLED
        self.mobility = False
        self.device_states = {}
        self.start_times = {}
        self.config = {}
        self.other_cfgs = {}
        self.device_states[name] = DeviceState.OFF
        
        # TODO: Setup all relevant publishers, subscribers, services, clients, etc
        self.system_state_sub = self.create_subscription(SystemStateMsg, "/autonav/shared/system", self.system_state_callback, 10)
        self.device_state_sub = self.create_subscription(DeviceStateMsg, "/autonav/shared/device", self.device_state_callback, 10)
        self.system_state_pub = self.create_publisher(SystemStateMsg, "/autonav/shared/system", 10)
        self.device_state_pub = self.create_publisher(DeviceStateMsg, "/autonav/shared/device", 10)
        self.config_sub = self.create_subscription(ConfigurationUpdate, "/autonav/shared/config/updates", self.config_callback, 10)
        self.config_pub = self.create_publisher(ConfigurationUpdate, "/autonav/shared/config/updates", 10)
        self.config_broadcast_sub = self.create_subscription(ConfigurationBroadcast, "/autonav/shared/config/requests", self.config_broadcast_callback, 10)
        self.config_broadcast_pub = self.create_publisher(ConfigurationBroadcast, "/autonav/shared/config/requests", 10)

        self.performance_pub = self.create_publisher(Performance, "/autonav/shared/performance", 10)
        self.log_pub = self.create_publisher(Log, "/autonav/shared/log", 10)

    def init(self) -> None:
        """
        Called when the node synchronizes with the system.
        """
        pass

    def _smart_dump_config(self, config) -> str:
        if isinstance(config, dict):
            return json.dumps(config)
        else:
            try:
                return json.dumps(config.__dict__)
            except:
                return str(config)

    def config_callback(self, msg: ConfigurationUpdate) -> None:
        """
        Callback for the configuration update topic.
        """
        if msg.device == self.get_name():
            # self.log(f"Received update on our own configuration", LogLevel.DEBUG)
            old_cfg = self.config
            new_cfg = json.loads(msg.json)
            self.apply_config(new_cfg)
            self.on_config_update(old_cfg, self.config)
        else:
            # self.log(f"Received updated on {msg.device}'s configuration", LogLevel.DEBUG)
            pass
        
        self.other_cfgs[msg.device] = json.loads(msg.json)

    def apply_config(self, config) -> None:
        """
        Apply a configuration to the node.
        Override this to use type hints for the config.
        """
        self.config = config

    def on_config_update(self, old_cfg, new_cfg) -> None:
        """
        Called when the configuration is updated.
        """
        pass

    def config_broadcast_callback(self, msg: ConfigurationBroadcast) -> None:
        """
        Callback for the configuration broadcast topic.
        """
        if msg.target_nodes == [] or self.get_name() in msg.target_nodes:
            self.broadcast_config()

    def _broadcast_config(self, name, config) -> None:
        """
        Broadcast our configuration to the system.
        """
        msg = ConfigurationUpdate()
        msg.device = name
        msg.json = self._smart_dump_config(config)
        self.config_pub.publish(msg)

    def broadcast_config(self) -> None:
        """
        Broadcast our configuration to the system.
        """
        self._broadcast_config(self.get_name(), self.config)

    def request_all_configs(self) -> None:
        """
        Request the configuration of all devices.
        """
        self.request_config(None)

    def request_config(self, device: str) -> None:
        """
        Request the configuration of a specific device.
        """
        msg = ConfigurationBroadcast()
        if device != None:
            msg.target_nodes = [device]
        else:
            msg.target_nodes = []
        self.config_broadcast_pub.publish(msg)

    def write_config(self, config) -> None:
        """
        Register a configuration object.
        """
        self.config = config
        self.broadcast_config()

    def perf_start(self, name: str) -> None:
        """
        Start a performance measurement.
        """
        self.start_times[name] = time.time_ns() // 1_000_000

    def perf_stop(self, name: str, print_to_console: bool = False) -> None:
        """
        End a performance measurement.
        """
        if name not in self.start_times:
            self.log(f"Performance tiemr {name} not found", LogLevel.ERROR)
            return
        
        duration = (time.time_ns() // 1_000_000) - self.start_times[name]
        self.start_times.pop(name)
        
        msg = Performance()
        msg.name = name
        msg.elapsed = duration
        self.performance_pub.publish(msg)
        
        if print_to_console:
            self.log(f"Performance timer {name} took {duration}ms", LogLevel.DEBUG)

    def system_state_callback(self, msg: SystemStateMsg) -> None:
        """
        Callback for the system state topic.
        """
        old_state = self.system_state
        old_mobility = self.mobility

        self.system_state = SystemState(msg.state)
        self.mobility = msg.mobility

        if old_state != self.system_state:
            self.on_system_state_updated(old_state, self.system_state)

        if old_mobility != self.mobility:
            self.on_mobility_updated(old_mobility, self.mobility)

    def on_system_state_updated(self, old: SystemState, new: SystemState) -> None:
        """
        Called when the system state is updated.
        """
        pass

    def on_mobility_updated(self, old: bool, new: bool) -> None:
        """
        Called when the mobility state is updated.
        """
        pass
        
    def device_state_callback(self, msg: DeviceStateMsg) -> None:
        """
        Callback for the device state topic.
        """
        old_state = self.device_states[msg.device] if msg.device in self.device_states else None
        self.device_states[msg.device] = DeviceState(msg.state)
        if (old_state == None or old_state == DeviceState.OFF) and DeviceState(msg.state) == DeviceState.WARMING and msg.device == self.get_name():
            self.init()

    def set_device_state(self, state: DeviceState) -> None:
        """
        Set the state of the current node.
        """
        self._set_device_state(self.get_name(), state)

    def _set_device_state(self, device: str, state: DeviceState) -> None:
        """
        Set the state of a specific device.
        """
        msg = DeviceStateMsg()
        msg.device = device
        msg.state = state
        self.device_state_pub.publish(msg)

        
    def set_system_state(self, state: SystemState) -> None:
        """
        Set the state of the system.
        """
        msg = SystemStateMsg()
        msg.state = state
        msg.mobility = self.mobility
        self.system_state_pub.publish(msg)

    def set_mobility(self, mobility: bool) -> None:
        """
        Set the mobility of the system.
        """
        self.mobility = mobility
        self.set_system_state(self.system_state)

    def get_device_state(self, device: str = None) -> DeviceState:
        """
        Get the state of a specific device.

        If no device is provided, the state of the current node is returned.
        """
        if device is None:
            device = self.get_name()

        if device not in self.device_states:
            return DeviceState.UNKNOWN
        
        return self.device_states[device]
    
    def get_system_state(self) -> SystemState:
        """
        Get the state of the system.
        """
        return self.system_state
    
    def is_mobility(self) -> bool:
        """
        Check if the system is in mobility mode.
        """
        return self.mobility

    def log(self, message: str, level: LogLevel = LogLevel.INFO) -> None:
        """
        Log a message with a given log level.
        """
        # Get the current time as yyyy-mm-dd hh:mm:ss:ms
        current_time = time.strftime("%Y-%m-%d %H:%M:%S:", time.localtime()) + str(time.time() % 1)[2:5]
        
        # Get the calling function name and line number
        frame = inspect.currentframe().f_back
        calling_function = frame.f_code.co_name
        line_number = frame.f_lineno
        
        # Get the log level as a string
        level_str = LogLevel(level).name
        level_str = level_str + " " * (5 - len(level_str))

        # Log to topic
        msg = Log()
        msg.timestamp = time.time_ns() // 1_000_000
        msg.level = level.value
        msg.node = self.get_name()
        msg.function_caller = calling_function
        msg.line_number = line_number
        msg.message = message
        self.log_pub.publish(msg)

        r, g, b = 0, 0, 0
        match level:
            case LogLevel.DEBUG:
                r, g, b = 61, 117, 157
            case LogLevel.INFO:
                r, g, b = 255, 255, 255
            case LogLevel.WARN:
                r, g, b = 226, 174, 47
            case LogLevel.ERROR:
                r, g, b = 195, 59, 91

        # FATAL requires backgroud colors so it gets its own if statement
        if level == LogLevel.FATAL:
            print(f"{sty.fg(99, 150, 79)}{current_time} {sty.fg.white}| {sty.bg(207, 62, 97)}{level_str}{sty.bg.rs} {sty.fg.white}| {sty.fg(90, 60, 146)}{self.get_name()}{sty.fg.white}:{sty.fg(90, 60, 146)}{calling_function}{sty.fg.white}:{sty.fg(90, 60, 146)}{line_number} {sty.fg.white}- {sty.bg(207, 62, 97)}{message}{sty.rs.all}")
            return
        
        print(f"{sty.fg(99, 150, 79)}{current_time} {sty.fg.white}| {sty.fg(r, g, b)}{level_str} {sty.fg.white}| {sty.fg(90, 60, 146)}{self.get_name()}{sty.fg.white}:{sty.fg(90, 60, 146)}{calling_function}{sty.fg.white}:{sty.fg(90, 60, 146)}{line_number} {sty.fg.white}- {sty.fg(r, g, b)}{message}{sty.rs.all}")

# Run Nodes currently don't exist here or anywhere else.. Pasting frm lst yrs code
    def run_node(node):
        """
        Runs the node with the correct ROS parameters and specifications

        :param node: The node to run.
        """

        executor = MultiThreadedExecutor()
        executor.add_node(node)
        executor.spin()
        executor.remove_node(node)

    def run_nodes(nodes):
        """
        Runs the nodes with the correct ROS parameters and specifications

        :param nodes: The nodes to run.
        """

        executor = MultiThreadedExecutor()
        for node in nodes:
            executor.add_node(node)
        executor.spin()
        for node in nodes:
            executor.remove_node(node)